### PR TITLE
Update Go benchmarks to use the Go testing harness

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Thanks to @kortschak

**Comment by btracey** [comment]
Sorry, I think the behavior of quicksort isn't correct here. I'll make an edit. 

That said, the quicksort benchmark right now is much more of a test of random number generation than a test of recursion/swapping. Right now, the random number generation is roughly 90% of the runtime. It would be better to have some really simple random number generator that all of the languages can use (good random numbers aren't needed, just the same unsorted list). Inside the quicksort loop, copy those values and sort. Copy is much quicker than generating random numbers, so it would be a better test.

**Comment by btracey** [comment]
Okay, behavior fixed. It's now apples to apples.

**Comment by ScottPJones** [comment]
The parse_int test is also really just a test of random number generation :disappointed:

**Comment by jiahao** [comment]
@btracey that's a rather peculiar observation about `quicksort` taking 90% of the time to generate the random vector. In contrast, Julia takes 5% of the time to generate the random input and 95% to do the sorting. Does Go have a canonical implementation of dSFMT? That is what we use in Julia by default.

@ScottPJones I just profiled the Julia version of the `parseint` microbenchmark and the two major components are `hex` (38%) and `parseint` (52%). The random number generation takes 0.6% of the run time.

**Comment by StefanKarpinski** [comment]
It may also be possible to call dSFMT from Go as a C library – I assume there's some kind of C FFI.

**Comment by btracey** [comment]
Sorry, I double checked and it's not 90% of the time (must have had something wrong before). Seems it's about 15% of the run time.

**Comment by jiahao** [comment]
OK good to know. If it was 90% of the runtime then I think it would have been worth triaging.

I can confirm that this PR runs on our test machine.

**Comment by jiahao** [comment]
Thanks!

File: test/perf/micro/perf.go [hunk 1]
```diff
@@ -194,106 +196,109 @@ func print_perf(name string, time float64) {
...
+		},
+	},
+
+	{
+		name: "parse_int",
+		fn: func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				for k := 0; k < 1000; k++ {
+					n := rnd.Uint32()
+					m, _ := strconv.ParseUint(fmt.Sprintf("%x", n), 16, 32)
```
Comment (kortschak) [inline test/perf/micro/perf.go]: Replace this with

```
m, _ := strconv.ParseUint(strconv.FormatUint(uint64(n), 16), 16, 32)
```

File: test/perf/micro/perf.go [hunk 2]
```diff
@@ -194,106 +196,109 @@ func print_perf(name string, time float64) {
...
-		assert(c.At(0, 0) >= 0)
-		d := float64(time.Since(t).Seconds())
-		if d < tmin {
-			tmin = d
-		}
-	}
-	print_perf("rand_mat_mul", tmin)
+func runBenchmarkFor(fn func(*testing.B)) (seconds float64) {
+	bm := testing.Benchmark(fn)
+	return bm.T.Seconds() / float64(bm.N)
```
Comment (kortschak) [inline test/perf/micro/perf.go]: Add

```
func main() {
	for _, bm := range benchmarks {
		seconds, err := runBenchmarkFor(bm.fn)
		if err != nil {
			log.Fatalf("%s %s", bm.name, err)
		}
		print_perf(bm.name, seconds)
	}
}

func runBenchmarkFor(fn func(*testing.B)) (seconds float64, err error) {
	bm := testing.Benchmark(fn)
	if (bm == testing.BenchmarkResult{}) {
		return 0, errors.New("failed")
	}
	return bm.T.Seconds() / float64(bm.N), nil
}
```

**Pull request merged**